### ISSUE TITLE
Support Math — somehow (general discussion of approaches and status)

### ISSUE THREAD

**Issue by akavel** [sile-typesetter/sile#220]
I'm aware MathJax is listed [on the wiki](https://github.com/simoncozens/sile/wiki/easy-jobs) as ["up-for-grabs"](http://up-for-grabs.net/), but I just wanted to start by asking: have you maybe tried pushing some [stuff from the MathJax's _unpacked/_ subdir](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/input/TeX/jax.js) through some [js-to-lua](https://github.com/tessel/colony-compiler/blob/master/src/colonize.js) translator ([or other](https://github.com/wizzard0/js2lua))?

I'm on Windows, so personally, before I maybe start trying to contribute, I have to first manage to compile the stuff (yes, I've seen the issue tracking this)...

_(By the way, I got here by accidentally watching the 2015 FOSDEM video of yours; very entertaining and interesting :) although I've apparently already +1'ed your project at some point, only this video made me aware of many things about it (both feature demos, architecture, and how it relates and improves over *TeX) and thus more interested.)_

-----

*edit:* A few years later, after some prototyping and experiments, I'm expanding this initial post with more detailed current status of the project. The path I've eventually picked to try and implement this is based on ASCIIMath, not MathJax.

## Current TODO tasks

Roughly in priority/painfulness order. I'd be especially grateful if someone could help with the ones marked "HELP NEEDED", I don't know how to proceed with those unfortunately :(. Jumping in to help with *any* task is enthusiastically encouraged!

   - [x] ~~Translate [ASCIIMathML.js](https://github.com/asciimath/asciimathml) from JS to Lua~~ *([done and tested](https://github.com/akavel/asciimath))*
   - [x] ~~Translate [SVGMath](https://github.com/w3c/svgwg/tree/73855c2f16b36122cfd3e917ceb4c7d8f6314834/tools/math) from Python to Lua~~ *([done](https://github.com/akavel/svgmath) with help of https://github.com/akavel/pylua based on https://github.com/kevinw/pylua)*
   - [x] ~~Write SVG parser and renderer for SILE supporting SVGs generated by SVGMath~~ *([done](https://github.com/simoncozens/sile/issues/468))*
   - [x] ~~Integrate the tools into a pipeline callable from SILE with an `\asciimath` directive~~ *([done](https://github.com/akavel/silemath))*
   - [x] ~~The `<text>` nodes in SVGs generated by svgmath have `text-anchor="middle"`, so their **horizontal position should be tweaked**, but I couldn't yet find out how to do that properly (moving them by `-shape[1].width/2` didn't work for me for now, it broke things more than fixed). This would be **important fix to the current buggy layout**. This should be verified by rendering some correct SVG (e.g. https://github.com/akavel/svgmath/blob/master/testdata/test13.svg) through the final part of the pipeline, that is, `\mathsvg`.~~ *([done](https://github.com/akavel/silemath/commit/b35d9d19266b0aaedb40661fa8854014c819dbbb))*
   - [ ] Make the **SILE bounding box** have correct size (how?), matching the content inside.
     - :sos: **\[HELP NEEDED\]** :sos: **how to force SILE to draw a debugging rectangle around the current bounding box?**
   - [ ] **Match the sizes of fonts** used by MathML->SVG translation to the font size used in the current paragraph in SILE
   - [x] ~~Use **TTF font paths from SILE** instead of hard-coded paths in [svgmath/config.lua](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/svgmath/config.lua#L20-L23)~~
     - done via https://github.com/akavel/silemath/commit/ef4dc29ccfdf2f1dfa5e398fb2497ffe8e266f07 &mdash; paths to TTF fonts are now taken from SILE. You should be able to just run sample.sil immediately after cloning the https://github.com/akavel/silemath repo &mdash; please try to test it and report results!
   - [ ] Write a README in akavel/silemath repo, with:
     - [ ] syntax examples & reference & BNF
     - [ ] result examples
     - [ ] installation guide &mdash; just cloning the https://github.com/akavel/silemath/ repo and running `sile sample.sil` should work now (after [recent commit](https://github.com/akavel/silemath/commit/ef4dc29ccfdf2f1dfa5e398fb2497ffe8e266f07))
   - [ ] Dynamically calculate font metrics for MathML->SVG translation, based on info from SILE, instead of current hard-coded parameters in [svgmath/config.lua](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/svgmath/config.lua)
   - [ ] Make it possible to choose the font(s) used by MathML->SVG translation
   - [ ] Fix bugs in rendering of some expressions; e.g. [some ASCIIMath samples](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/sample.sil#L32-L42) currently crash later stages of the pipeline (IIRC the MathML->SVG translation).
   - [ ] Fix any remaining bugs in SVG rendering step. This should be verified by rendering some correct SVG (e.g. https://github.com/akavel/svgmath/blob/master/testdata/test13.svg) through the final part of the pipeline, that is, `\mathsvg`.
   - [ ] Fix layouting bugs in svgmath
     - Note: should fix `\mathsvg` first, as mentioned in the task above
     - There are a lot of testcases in https://github.com/akavel/svgmath/tree/master/testdata. The *.out.svg files were generated by Lua, and the *.svg files by Python. (1) Fix the Lua code so that the *.out.svg files match the *.svg files. (2) Fix the few files which crashed the original Python code too (e.g. test20.mml?).
     - consult https://github.com/acivux/svgmath, maybe they fixed some of the bugs present in the original Python code
   - [ ] Integrate into SILE core — consult with @simoncozens
   - [ ] \[LATER\] Write some smarter parser on SILE side, so that:
     - users don't have to escape each `{` and `}` and `\` in math expressions
     - maybe "common" LaTeX syntax like `$ ...math... $` or `\[ ...math... \]` could be used?
   - [ ] :sos:  **\[HELP NEEDED\]** :sos: simplify the [*renderGlyph* function](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/mathsvg.lua#L11-L28) — currently it looks super ugly and probably very non-idiomatic SILE code, but I have no idea if/how it could be improved
   - [ ] \[OPTIONAL\] Migrate above tasks to issues in the akavel/silemath repo.
   - [ ] \[LATER\] Add support for UnicodeMath, which seems to be an interesting new development, inspired by ASCIIMath, apparently supported by both [MS Office/Word](https://support.office.com/en-us/article/linear-format-equations-using-unicodemath-and-latex-in-word-2e00618d-b1fd-49d8-8cb4-8d17f25754f8) and [LaTeX](https://ctan.org/pkg/unicode-math) (TODO: or is it just a name collision??), and thus apparently used by some people already. (See also [some more or less related plugin for SublimeText](https://github.com/mvoidex/UnicodeMath).) 

### Current example result

<img src="https://user-images.githubusercontent.com/273837/36398082-a5775514-15c5-11e8-8478-66697ccada50.png" width="400"/>

**Comment by akavel** [comment]
Notes from the **first failed attempt**: I tried translating [KaTeX](https://github.com/Khan/KaTeX#server-side-rendering-or-rendering-to-a-string) with [castl](https://github.com/PaulBernier/castl). The translation "succeeded" (using KaTeX v0.7.1 from git and castl 1.2.4), with commands roughly as below:

```
$ cd katex
$ git checkout v0.7.1
$ for f in katex.js src/*.js; do castl $f -o --debug --node; done
```

However, trying to run the following Lua script afterwards with Lua 5.2:

```
local katex = assert(loadfile 'katex.js.lua')()
```

failed with a Lua error: `[...] too many C levels (limit is 200) [...]`, which [seems to mean there's too deep nesting in the code](http://lua-users.org/lists/lua-l/2013-03/msg00061.html).

**Comment by akavel** [comment]
After some more looking around, I want to try my luck with ASCIIMath. It seems small enough, that even in case js2lua converters won't help, a manual port seems feasible. It does not implement Full Parity With TeX&trade;; but as we all know too well I think, TeX is Turing-complete, and full parity is not necessarily even desirable. Anyway, even if I succeed, in Future&trade; someone can always just ditch the ASCIIMath and/or incrementally replace it with something More Powerful&trade; or whatever.

Now, with that said, there's one thing I now realized I'm not really clear on. Namely, **what** would be **the best format to output** from a math module **for SILE to consume?** ASCIIMath I think produces MathML; KaTeX claims to emit HTML+CSS; MathJaX is I believe super-duper-dynamic-adaptive-autodetect and can produce MathML, SVG, possibly images, and I don't even know what else. What would SILE like best? Imagine we're free of constraints; would it be **PDF? SVG? Lua rendering code? glyphs + coords + scales? PostScript?** glues penalties and some other weird stuff like this? (not even sure if it's there for math). On the other hand: is there/are there some *simple* format(s) (hopefully, vector) that SILE could reasonably accept for a reasonable first prototype/PoC?

As a side note for future: in case of failure, I think one more direction to research would be troff's *eqn* processor.

**Comment by suhr** [comment]
> I want to try my luck with ASCIIMath.

It doesn't have `\cfact` though, not even talking about commutative diagrams.

Anyway, good luck. It is hard to live without even basic math.

**Comment by alerque** [comment]
PDF would seem to make the most sense. That way the output can be directly embeded without much chance for scale or other interpretive errors.

An SVG route would be an entirely different pipeline and introduce some other pain points, but if there was a Math→SVG processor we might be able to bolt that on entirely outside of Lua.

**Comment by akavel** [comment]
@suhr Did you ever try [groff's eqn](https://github.com/jgm/pandoc/issues/1839)? Do you think it's currently more powerful or less, compared to ASCIIMath?

If PDF sounds best, then I think eqn sounds like more bang for the buck for a first shot. With various caveats, first among them GPL I suppose (or alternatively GPL-incompatible Lucene license for Plan9's troff+eqn; not sure about PDF support there too). Or do you know by chance of someone already trying to integrate eqn with SILE and stumbling upon some problems?

**Comment by akavel** [comment]
**Prototype translation of [ASCIIMathML.js](https://github.com/asciimath/asciimathml) to Lua [is done](https://github.com/akavel/asciimath) and tested**. I used the [very cool castl translator](https://github.com/PaulBernier/castl). Also, given that it's not my first port, I must shout out, that [comprehensive end-to-end tests](https://github.com/asciimath/asciimathml/blob/dd021360fbcfd5f5eaa7f4711b3a662a502db409/test/unittests.js) do *really* boost porting speed *awesomely!* 

TODO: test if [svgmath](https://github.com/w3c/svgwg/tree/73855c2f16b36122cfd3e917ceb4c7d8f6314834/tools/math) will work reasonably well to translate the resulting MathML to SVG. It seems quite a small and readable MIT codebase, so should be feasible to hand-port to Lua if it's working well. I imagine it should be possible to migrate the SVG emitter to a PDF-or-whatever emitter more fitting for SILE later, should the PoC prove successful.

**Comment by alerque** [comment]
@akavel SILE already has a way to embed PDFs inside PDFs. Use the `\img[src=file.pdf]` function passing it a PDF as the input image. This is easy and reliable because the PDF doesn't have to be edited in any way, it can basically just be wrapped up and packaged inside the final output PDF with a location where it should be included.

Yes there is also rudimentary SVG support, but this isn't really up to snuff yet and involves being able to render whatever is in the SVG and convert it to a post script representation that can be embedded in the PDF. This is nowhere near as reliable as just nesting PDF content. It works alright for some things, but doesn't understand a lot of SVG instructions yet.

**Comment by akavel** [comment]
@alerque Do you know if it's possible to create a PDF with "tiny paper size", so that it would have size of exactly the bounding box of an equation (i.e. 2cm x 1cm, or something like that)? Or does `\img[src=...]` crop the contents to extract only the "black" contents, ignoring paper size?

**Comment by alerque** [comment]
Embeding a PDF as an image does nothing in the way of cropping or finding contents, it just uses the bounding box and inserts that at the requested overall width, hence scaling the source's paper size to the embeded size.

Obviously if you're trying to bold on some other render you might end up with an awkward mating like this, but ideally of course a native Lua parser would transform the formula to native drawables that could be passed to the output engine (which at this point is  basically always the PDF output engine but keep in mind that isn't supported to be a hard coded limitation).

**Comment by akavel** [comment]
Ok, did a quick test with groff and indeed this direction seems to (kinda) work already. Two main caveats:
- groff's "document-class" counterpart is obviously unaccustomed to tiny paper sizes, pushing the equation to second page if I reduce "page size" to 1in x 1in, making it not show up in SILE. Learning more groff/troff (or finding someone with knowledge) would be needed to adjust/build an appropriate "document class".
- how to get information about the baseline in the original PDF?

For anyone interested, here comes a quick & rough transcript of my session. Note I'm using [Nix package manager](http://nixos.org/nix):
```
$ nix-channel --list
nixos http://nixos.org/channels/nixos-17.03

$ nix-shell -p sile groff

[nix-shell:~]$ cat try3.troff    # NOTE: encoded in UTF-8, as expected for `-k` below IIUC
.PRINTSTYLE TYPESET
.PAGE 1i 2i .05i .05i .05i .05i
.FAMILY T \" Times Roman
.START
.EQ
∇²ϕ = 0
.EN

[nix-shell:~]$ cat eqn-pdf.sil
\begin[class=book,papersize=129mm x 198mm]{document}
\script[src=packages/image]
...
Foo bar \img[src=try3.pdf] baz.
...
\end{document}

[nix-shell:~]$ pdfmom -e -k try3.troff > try3.pdf
[mom]: Your chosen bottom margin for running text is too close to the footer margin.
       No footers or bottom-of-page page numbers will be printed.
       Please reset B_MARGIN or FOOTER_MARGIN to allow enough space.
       If no footers or bottom-of-page page numbers are required,
       invoke .FOOTER_MARGIN 0 before .START
troff: environment stack underflow

[nix-shell:~]$ sile eqn-pdf.sil
This is SILE 0.9.4
</vm-share/eqn-pdf.sil>

! Font 'Gentium' not available, falling back to 'DejaVu Sans'
[1] [2] 
```

[screenshot pending...]

**Comment by akavel** [comment]
For the record, a "more advanced" groff equation worked as well:
```
sum from i=0 to {inf} x sup i { sqrt {alpha + beta} } over {sin (x)}
```
[screenshot pending]

**Comment by alerque** [comment]
> how to get information about the baseline in the original PDF?

What do you mean by "original" here? The baseline in the SILE PDF should be straight forward enough. The trick is going to be leaning where the baseline of an equation PDF is in the event of an inline equation such as the ones in your examples. Once you do learn that you can offset the image in  SILE using `\lower[height=Y]{}`. Combined with the right leading mechanism (`fit-glyph`?) this should allow lines to move around to accommodate formulas.

**Comment by akavel** [comment]
(@alerque Yes, I meant exactly what you mean by "learning where the baseline of an equation PDF is".)

Personally, at this point I still plan to pursue the path of: ASCIIMath &rarr; MathML &rarr; SVG &rarr; SILE. I believe it's more promising in the long run. That said, groff (specifically, eqn + pic) seems totally more useful in the short run, especially given that their language is much more powerful and expressive than ASCIIMath (I imagine commutative diagrams should be totally doable with eqn or eqn+pic). But as far as my current plans go, I'm leaving perfecting of the groff path as an... *cue ominous thunder...* "excercise for the reader".

**Comment by akavel** [comment]
Cool, [SVGMath](https://github.com/w3c/svgwg/tree/73855c2f16b36122cfd3e917ceb4c7d8f6314834/tools/math) seems to work! Below is a quick screenshot of: 
```
sum_(i=0)^oo x^i (sqrt(alpha + beta))/sin(x)
```
after passing through:

&nbsp; `asciimath.lua` &mdash;*(MathML)*&rarr; `math2svg.py` &mdash;*(SVG)*&rarr; `Firefox`

(Note: I had to add `<html><head></head><body>` boilerplate + MathML namespace on the `<math>` element.)

The screenshot:

![screenshot](https://cloud.githubusercontent.com/assets/273837/25091244/5b200fd6-2388-11e7-82be-fba4303e2c26.png)

**Comment by simoncozens** [comment]
I don't have much to add to this but I'm really happy it's being worked on. The baseline/leading issues will be quite easy to fix when then time comes.

**Comment by akavel** [comment]
@simoncozens How about the bounding box? If groff emitted a big, white page with a (comparably) small equation in some quasi-random place on it, would you find it easy to extract just the "dirty" area? If so, it would make the task *much* easier on the groff side, I think &mdash; it should be just deleting all page decorations, and *voil&agrave;,* the thing's done?

**Comment by alerque** [comment]
@akavel I don't think SILE should be responsible for parsing PDF data. It really isn't setup for that. It knows how to write it and it can take a chunk of PDF data and embed it somewhere, but parsing through the PDF data and doing a virtual rendering to find the bounds of actual content in it is more than a little outside the purview of the typesetting engine.

That being said it  may not have to be `groff` that does the work. There are other tools for examining and modifying PDFs. Given that `groff` is already an external dependency it wouldn't be entirely out of the question to include another  tool in the pipeline. However I'm concerned that even if we found such a tool the result would still be a hack.  What happens when the formula is so large that it overflows the random spacious page size you assign? Is there really no way to have `groff` match the CropBox value to the resulting content it generates?

**Comment by akavel** [comment]
*(@alerque Sorry, but can you please remove my "assignee" status from this issue? I don't want this to start feeling like work to me; I'm doing this just out of fun and because I want to help and like this project. A subjectively perceived connotation of "responsibility" and "obligation" coming from the assignment status unfortunately heavily impacts on my fun and is counter-productive for me working on this. Also, I don't want to make anyone feel like "this issue is occupied so I won't waste my time contributing to it"; I'm totally ok with someone else even finishing it before me and closing the issue (I don't lack things to do). Finally, I never promised I'll complete this, so I don't want anybody to think so by accident. I just fancy to have a stupid plan to try my luck at this, with 100% undefined schedule. As also shown by a ~1.5yr gap between my first and second comment in this thread. TIA!)*

**Comment by alerque** [comment]
@akavel Of course. I didn't intend it in the sense of responsibility at all only that because you've got code kicking around in this department it would make sense to coordinate with you before launching into this one.

Speaking of which do you have a branch somewhere with your work in progress? If it was posted and linked from here that would give somebody else the chance to come along and extend or clean it up.

**Comment by akavel** [comment]
@alerque Thanks!

Re: branch link &mdash; sure, good idea. Summarizing:

- [x] The asciimath port is already done, see: https://github.com/akavel/asciimath.
- [ ] Now I'm working on porting svgmath (a Python MathML->SVG renderer/translator) to Lua here: https://github.com/akavel/svgmath. 
  - To help with this I'm using a fork of kevinw's pylua &mdash; a basic Python->Lua translator &mdash; to automate as much as possible, extending it as needed here: https://github.com/akavel/pylua.
- [ ] Later, some form of integration with SILE will be needed, but I haven't forked SILE yet, too early still.

As an alternative, `groff` could be used, as [mentioned in a comment above](https://github.com/simoncozens/sile/issues/220#issuecomment-294478476). This would also require integration with SILE, and some more research on `groff` side how to lay out the PDFs to make importing them into SILE easier (or, alternatively, do some processing on SILE side).

**Comment by akavel** [comment]
The [Lua port of SVGMath](https://github.com/akavel/svgmath) is now giving results, and very similar ones to what the original Python version gave. On the [available test data/examples (*.mml)](https://github.com/akavel/svgmath/tree/master/testdata), some have minor/medium spacing issues, and there are some non-italic &pi; characters. Also, 2 examples fail, but they fail in Python too (test18 and test20). 

## Example Lua results (testdata/*.out.svg):
> [![](https://cdn.rawgit.com/akavel/svgmath/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test23.out.svg)](https://github.com/akavel/svgmath/blob/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test23.out.svg)
> ---
> [![](https://cdn.rawgit.com/akavel/svgmath/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test19.out.svg)](https://github.com/akavel/svgmath/blob/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test19.out.svg)


## Reference Python results (testdata/*.svg):
> [![](https://cdn.rawgit.com/akavel/svgmath/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test23.svg)](https://github.com/akavel/svgmath/blob/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test23.svg)
> ---
> [![](https://cdn.rawgit.com/akavel/svgmath/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test19.svg)](https://github.com/akavel/svgmath/blob/6cf404d5b9d475a66ecee1bbb87225d4dfdd53e2/testdata/test19.svg)




At this point I'd like to ask for help integrating this as an experimental plugin/feature into SILE. Can you possibly give me some guidance on how to start? I'd be very grateful for hints, would considerably speed this up for me. Please note, that I have no practical knowledge/experience of SILE, other than reading the manual.

**Comment by akavel** [comment]
Uh, oh; do I see correctly that SVG import seems not yet working as of current state of the SILE repo?

**Comment by simoncozens** [comment]
Try it... We use the nanosvg parser to render SVG to a set of paths, and then there's code in `src/svg.c` that turns it into PDF drawing operators. There may be incompleteness in either of these steps, but if there is, file a bug and I'll work on adding support.

Look in `packages/svg.lua` for an SVG package which loads in a .svg file. If you have a string containing SVG instead of a file, you will need to add another command to `svg.lua` based on `\include-svg-file`.

**Comment by akavel** [comment]
@simoncozens Ok, thanks. I've reported #468. My initial suspicion is that maybe nanosvg ignores SVG's `<text>` elements... :/

**Comment by simoncozens** [comment]
OK. Well, I would reconsider the use of nanosvg, if it's not actually going to help us render real-world SVGs. If we want to do this *right* (which we do) then I guess we need to be looking at librsvg or cairo or something.

**Comment by akavel** [comment]
Alternatively, we could try extending nanosvg to support `<text>`? Or is it too much of a stretch? If that's the only missing thing (? no idea), maybe it wouldn't be very stupid? [Also, shouldn't this part of the discussion better move to #468?]

**Comment by akavel** [comment]
Tried pushing the SVG through `inkscape -z -f test13.svg --export-pdf test13.pdf`, then doing `\img[src=test13.pdf]`. Seems to "work", though font sizes in the equation are totally inadequate w.r.t. surrounding text (kinda 2x bigger; truly *display*math...)

**Comment by akavel** [comment]
svgmath seems to be producing only a limited set of SVG primitives, namely: path, rect, line, `<g translate=...>`, text, circle. I think I could just parse the "SVG" by hand in Lua, reserve a properly sized hbox, then try emitting appropriate SILE/pdf primitives, probably on `outputYourself`. (Where) could I find functions to emit appropriate primitives in SILE? or would it be in libtexpdf?

**Comment by simoncozens** [comment]
This should do what you want:

```lua
local pdf = require("justenoughlibtexpdf")
...
pdf.add_content("q")
pdf.add_content(stuff)
pdf.add_content("Q")
```

**Comment by akavel** [comment]
Status update: the main integration step, or *leap*, is now done! 😃 

The following screenshot:

![](https://user-images.githubusercontent.com/273837/36281211-28075802-129d-11e8-919c-b31e2bbd99fa.png)

was rendered from the [following `sample.sil` file](https://github.com/akavel/silemath/blob/f7c1422c696df59b6f03ca3b065e1132bd2de2d1/sample.sil):

```
\begin{document}
Sample inital text.

Another line before.

\script[src=mathsvg]

Sample final text.

Sample more more more text.

\asciimath{E=mc^2}

----

Some examples from ASCIIMath homepage of Peter Jipsen:

- complex subscripts are bracketed, displayed under lim

. \asciimath{d/dxf(x)=lim_(h->0)(f(x+h)-f(x))/h}
\bigskip

- f^((n))(a) must be bracketed, else the numerator is only a

. \asciimath{f(x)=sum_(n=0)^oo(f^((n))(a))/(n!)(x-a)^n}
\bigskip

% FIXME(akavel): error
%- piecewise defined functions are based on matrix notation
%
%. \asciimath{x/x=\{(1,if x!=0),(text\{undefined\},if x=0):\}}
%\bigskip

% FIXME(akavel): error
%- angle brackets and invisible brackets
%
%. \asciimath{<< a,b >> and \{:(x,y),(u,v):\}}
%\bigskip

- symbols can be stacked

. \asciimath{stackrel"def"= or \\stackrel\{\\Delta\}\{=\}" "("or ":=)}
\bigskip

- prescripts simulated by subsuperscripts

. \asciimath{\{::\}_(\\ 92)^238U}
\bigskip

\end{document}
```

The full repo at its current state is at: https://github.com/akavel/silemath. If you try to use it, you may need to edit paths to TTF fonts on your system, currently hardcoded in [svgmath/config.lua, l.20-23](https://github.com/akavel/silemath/blob/f7c1422c696df59b6f03ca3b065e1132bd2de2d1/svgmath/config.lua#L20-L23).

**Comment by akavel** [comment]
Current TODO tasks:

   - [ ] The `<text>` nodes in SVGs generated by svgmath have `text-anchor="middle"`, so their **horizontal position should be tweaked**, but I couldn't yet find out how to do that properly (moving them by `-shape[1].width/2` didn't work for me for now, it broke things more than fixed). This would be **important fix to the current buggy layout**. This should be verified by rendering some correct SVG (e.g. https://github.com/akavel/svgmath/blob/master/testdata/test13.svg) through the final part of the pipeline, that is, `\mathsvg`.
      - is it possible to make SILE draw debugging rectangles around (or instead) of glyphs?
   - [ ] Fix any remaining bugs in SVG rendering step. This should be verified by rendering some correct SVG (e.g. https://github.com/akavel/svgmath/blob/master/testdata/test13.svg) through the final part of the pipeline, that is, `\mathsvg`.
   - [ ] Fix layouting bugs in svgmath
     - Note: should fix `\mathsvg` first, as mentioned in the task above
   - [ ] Make the **SILE bounding box** have correct size (how?), matching the content inside.
     - **how to force SILE to draw a debugging rectangle around the current bounding box?**
   - [ ] **Match the sizes of fonts** used by MathML->SVG translation to the font size used in the current paragraph in SILE
   - [ ] Make it possible to choose the font(s) used by MathML->SVG translation
   - [ ] Use **TTF font paths from SILE** instead of hard-coded paths in [svgmath/config.lua](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/svgmath/config.lua#L20-L23)
   - [ ] Dynamically calculate font metrics for MathML->SVG translation, based on info from SILE, instead of current hard-coded parameters in [svgmath/config.lua](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/svgmath/config.lua)
   - [ ] Fix bugs in rendering of some expressions; e.g. [some ASCIIMath samples](https://github.com/akavel/silemath/blob/f17699cd61fcda2a842c2a5b3ab83bbb89fc5a1d/sample.sil#L32-L42) currently crash later stages of the pipeline (IIRC the MathML->SVG translation).
   - [ ] \[OPTIONAL\] Migrate above tasks to issues in the akavel/silemath repo.
   - [ ] Write a README in akavel/silemath repo, with:
     - syntax examples & reference & BNF
     - result examples
     - installation guide
   - [ ] Integrate into SILE core — consult with @simoncozens
   - [ ] \[LATER\] Write some smarter parser on SILE side, so that:
     - users don't have to escape each `{` and `}` and `\` in math expressions
     - maybe "common" LaTeX syntax like `$ ...math... $` or `\[ ...math... \]` could be used?

**Comment by alerque** [comment]
Pro-tip: if you edit that TODO list into the first post in this thread Github has some features showing overall issue progress. Also especially for the held needed items I would recommend splitting them out into separate issues so they can be tackled and tracked independently, and just cross-linking the issue in the todo list for this issue. And no, I wouldn't recommend using a different repo for this since the goal is to have it all integrated into the sile core.

When I get get a chance I'll definitely have a look and see if there are any items I can help with.

**Comment by alerque** [comment]
Re: not having to escape `{` and similar characters in math functions: I already did most of the work needed on the LPEG parser to make this happen and it's in SILE core. I can help add handling for the math tags so they get the same treatment as Lua code now does. There will be one caveat for the math use case and that is that the _inline_ version of tags need balanced numbers of opening and closing braces if it stands a chance of finding the right closing mark for the tag. Using a "math mode" syntax ala LaTeX would be one way to alleviate that gotcha situation.

**Comment by akavel** [comment]
@alerque Regarding parsing, personally I was thinking it might be nice if this feature (i.e. some kind of "plugging in with new syntax into the parser") could be maybe available to any modules. My argument is such, that ASCIIMath is a controversial choice, and people might want to like to use some other engine in future, so there should ideally be some possibility for swapping them in the middle of a document. Also even ASCIIMath may incorporate some changes, resulting in different rendering of some expressions, so it would be nice to be able to version it (and pick a particular version of ASCIIMath in parts of a doc). On the other hand, there's a question if giving too much parser influence power to plugins could have bad consequences (see TeX...). So maybe just `$ ... $` and `\[ ... \]` could be the only special cases, and math engines would connect to some API hook in SILE to be able to capture/handle them?

**Comment by akavel** [comment]
Fixed another sub-issue; current example result:

<image src="https://user-images.githubusercontent.com/273837/36397415-fc572de4-15c2-11e8-80a2-7ad9036410b6.png" width="400"/>

**Comment by ctrlcctrlv** [comment]
I've read this issue and #578 (+856 SLOC!), and really don't understand why so much ... I don't know how to put it, perhaps, work? ... is being done on this, yet nothing in core since 2015.

#578 seems basically abandoned by @harrysummer, which is a shame, given how much time has obviously been put into it. This idea as well, doesn't seem to be going anywhere.

Also, this issue is misnamed. @akavel gave up on MathJax, instead opting for `ASCIIMath`.

So, I'd like to propose recalibrating. Early on here, KaTeX and MathJax were jettisoned because they couldn't be converted to Lua. I feel like nobody asked the obvious question: so what?

Node.js is available in most distros, making it an optional build dependency to me, quite honestly, seems like no big deal at all. After all, if lilypond (#435) and TiKZ (#437) are being considered, then I really don't think "can't be done in pure Lua" is a logical blocker.

So, my proposed solution would be, a `mathjax.lua` to call out to `node mathjax_wrapper.js`, which would actually import MathJax and set up command line options for controlling its output, outputting SVG onto stdout. Then you can just put `\mathjax{}` in your document using the #947 hack, and it uses the `svg` package to draw the math into your SILE document.

Now yes, this is less elegant, sure. But it has a key advantage of not being an enormous amount of work, meaning, it's actually likely to be finished by someone without superhuman stamina :-)

**Comment by akavel** [comment]
@ctrlcctrlv Those are the highs and lows of open-source 🙂 you are always welcome to try your idea and see if it works, and then try to contribute it back and see if the owner is OK with accepting it to the main codebase 🙂 and once you try, you might start to understand and grudgingly appreciate why this is as is 😆 and some day someone maybe will have enough patience and drive to push some version of it all the way to a successful conclusion.

**Comment by ctrlcctrlv** [comment]
@akavel Well sure, that's why I wrote the above, to spur discussion on what to do about this problem. 😄 

It's not a criticism of you, rather, I'm wondering, given that two highly passionate developers burned out on this project approaching it one way, whether or not it makes sense to come at the problem a different way.

**Comment by alerque** [comment]
I don't see any conflict with perusing multiple approaches and if either/both become usable then great. A more native approach using SILE to do the rendering of individual nodes would bring a lot of advantages, but is clearly the longer road to travel. A more modular approach of embedding a different language in the content of a tag and shipping it off to an external dependency to render an SVG and embedding the result is something we've shown to be workable too. The former does make more sense to attach ASCIIMath and/or TeX math compatibility, the latter makes more sense for MathJAX. Both could happily survive in parallel and serve different needs.

For the latter approach one SVG blob per command would be a good starting point. Bonus points would be an option that returned the resulting graphic broken up into lines that SILE could then use as frame break points.

Lets split up tracking issues though.

**Comment by alerque** [comment]
I haven't re-read through this issue and the todo list yet, but it needs an overhaul now that #1103 has been merged.

**Comment by akavel** [comment]
I just noticed the merge of #1103, I'm super happy and excited about it!!!!!!

From quickly glancing #1103, do I understand correctly it currently provides a MathML parser & translator, + a basic Tex-like parser? Do we know how much of MathML is supported, are there some/any limitations in the parser & translator?

If the MathML support is complete enough, I feel it should be pretty easy for me to cut this part out of my PR, and keep just the ASCIIMath->MathML translation part, thus maybe adding a new command `\asciimath` for those who might prefer that syntax. I'm really curious how much of it might be working with the present MathML renderer!!!! 🤩

**Comment by OlivierNicole** [comment]
I'm enthusiastic too!

The supported MathML tags are summarized in the checklist in #1103. Basic tags of the presentation markup are supported, but some are not supported (notably `mtext`), as well as some options (such as styling). No semantic tag is supported (but I believe that this part of MathML is much less used).

It should be pretty simple to define a command with a custom parser. You can look at the definition of the `math` and `mathml` commands in `packages/math/typesetter.lua`: the only difference between them is that `math` first converts its input into MathML.

**Comment by akavel** [comment]
Hm; @alerque is there some Windows nightly build I could download to try and work on this on MSWin? My main machine is Windows, so it'd be much easier for me this way; but if it's not possible, I'd be grateful if you could let me know and I'll try to undust my 2ndary NixOS laptop and see what I can do there...

**Comment by alerque** [comment]
@akavel The state of our Windows builds are in shambles. The CI runner needs somebody with expertise and interest in fixing it, it's been broken for a while. Nothing too major has changed architecturally since it was working, but I'm sure the Cmake definitions are well behind. I know there are quite a few people waiting for Windows builds and an installer, it is just so far outside the realm of my own production needs (and expertise) that it's hard for me to invest time it its upkeep.

I know NixOS works, in fact 0.11.1 is packaged there and I just used it for something (I'm looking at using NixOS to build the examples for the SILE website).

**Comment by akavel** [comment]
@alerque got it, thanks! IIUC I'll probably need to tweak the Nix expression to fetch from current master, not from 0.11.1, so that I could have the features that were needed in core for the mathml package to work, right?

**Comment by alerque** [comment]
@akavel True. Actually I would be really interested to know how to build a HEAD package for Nix too, but we should probably ask that on their issue tracker or look into other venues for how such a thing is done. Arch Linux has a `sile-git` package in the AUR which you can get running pretty easily. I can post a Dockerfile to get that running if that would help you.

Math will land in a tagged release soon, really the only hold up is wanting to have examples fixed on the website before we attract attention with the math release announcement.

**Comment by OlivierNicole** [comment]
I've been trying to compile HEAD again on Arch but ran into problems, as I posted on gitter.im.

**Comment by akavel** [comment]
@alerque as to building HEAD package, what I did and seemed to work was:

 1. copy the [default.nix from nixpkgs](https://github.com/NixOS/nixpkgs/blob/6f5e9a6f7bc5b83bcfd09f4a2d32e3a35d62fb16/pkgs/tools/typesetting/sile/default.nix) into local git repo with sile (found the path thanks to "[The NixOS Hound](https://search.nix.gsc.io/?q=sile%20%3D&i=nope&files=&repos=)")
 2. tweak the Nix expression's args list to work outside nixpkgs, by making it by default pull packages from `<nixpkgs>`:

        { nixpkgs ? import <nixpkgs> {}
        , lib ? nixpkgs.lib
        , stdenv ? nixpkgs.stdenv
        , darwin ? nixpkgs.darwin
        , fetchurl ? nixpkgs.fetchurl
        , makeWrapper ? nixpkgs.makeWrapper
        , pkg-config ? nixpkgs.pkg-config
        , poppler_utils ? nixpkgs.poppler_utils
        , gitMinimal ? nixpkgs.gitMinimal
        , harfbuzz ? nixpkgs.harfbuzz
        , icu ? nixpkgs.icu
        , fontconfig ? nixpkgs.fontconfig
        , lua ? nixpkgs.lua
        , libiconv ? nixpkgs.libiconv
        , makeFontsConf ? nixpkgs.makeFontsConf
        , gentium ? nixpkgs.gentium
        }:

 3. `$ nix-build` - worked for me!
 4. `$ ./result/bin/sile`

        SILE v0.11.1 (Lua 5.2)
        > 

Haven't tried to do anything more advanced with it yet.

(Please note I'm kinda early-intermediate with Nix, definitely not an expert.)

**Comment by suhr** [comment]
Don't forget to update the hash, otherwise nix might pull a wrong source.

**Comment by suhr** [comment]
Anyway, you want this if you want to build from master:

```
  src = fetchFromGitHub {
    owner = "sile-typesetter";
    repo = "sile";
    rev = "190683a82eabff2aeccc29fc36d3c1d610b70a87";
    sha256 = "sha256-vStmwsdqwD7yIdNYCk3Vu853fXhU661xFuRmzsa3N2M=";
  };
```

**Comment by alerque** [comment]
@akavel When you get that working I'd be happy to accept a PR to put it in the project source repo. I'm coming up with uses for it already, starting with the SILE website. Building the examples in CI is great but it produces a chicken and egg problem because the CI runner has the last tagged version. Or if there is an official way to setup the sile package in nixpkgs to have a HEAD build option (like Homebrew has) that would be great too. @suhr happen to know anything about that?

**Comment by suhr** [comment]
I would probably just add `flake.nix` to the project. Need fixing the build first though, and I'm too lazy to do that.

**Comment by akavel** [comment]
@alerque @suhr initial flake.nix attempt, up for discussion in the PR at: https://github.com/sile-typesetter/sile/pull/1223; managed to get it to work on my machine.

**Comment by akavel** [comment]
@OlivierNicole hmh; so, I tried, and generally, I managed to fairly quickly get able to pass the asciimath->mathml translator's output to the `\mathml` command. Thing is, it appears that it quite extensively uses some elements not yet supported by `\mathml`; notably, from what I already observed in some attempts:

 - `<mover>`
 - `<munder>`
 - `<munderover>`
 - `<mtext>`

A good (most probably ultimate) reference on various MathML outputs that ASCIIMath can produce can be seen at: https://github.com/akavel/silemath/blob/72102bde232f8d3828ad5fee67a3b268dccd872a/asciimath/testall.lua

So, the difficult question now: is there a chance you might consider trying to add support for those elements to the `\mathml` command?...

**Comment by akavel** [comment]
@OlivierNicole Alternatively, if I wanted to try my luck at attempting to add support for those elements, would you possibly fancy giving me some hints as to where to even start?

**Comment by OlivierNicole** [comment]
@akavel Regarding `<mover>`, `<munder>` and `<munderover>`, you are right that they are not implemented. Or rather, they are half implemented, because those tags are used for two things: typesetting operator limits and typesetting so-called “accents” and “under-accents”.

Accents are not supported yet, mainly because it requires to support the construction arbitrarily stretchable glyphs, something conceptually not complicated and all the pieces are there, I just did not prioritize that in the window of time when I was working on this.

Limits are supported. However, I did a mistake in my interpretation of the MathML standard, so limits are managed by the `<msup>` and `<msub>` tags. For instance, you can typeset a sum as `<msubsup><mo>&sum;</mo> <mrow>i = 0</mrow> <mn>n</mn></msubsup>`, whereas the spirit of MathML commands that you use `munderover` instead of `msubsup`.

I have unfortunately little time these days to work on this side project, but I can try to fix this and support big operator limits through `<munder>` and `<mover>`. The accents I can't work on yet, however.


> @OlivierNicole Alternatively, if I wanted to try my luck at attempting to add support for those elements, would you possibly fancy giving me some hints as to where in the code could I even start looking into this? I mean especially the math layouting engine, I think...

I believe the math package documentation in the manual (at the end of the section) gives a very brief overview of how to implement support for new MathML tags.

**Comment by akavel** [comment]
@OlivierNicole thanks! I now see [the lines near the end in packages/math.lua](https://github.com/OlivierNicole/sile/blob/22ec44442be56506cce8af50b5efbd05ff327654/packages/math.lua#L270-L278) describing `mbox`, `shape`, `output`, etc.

If you could possibly do the `<mover>`/`<munder>`/`<munderover>` fix, that would be awesome! Do you think things equivalent to `\stackrel` ([a.k.a. `\underset`/`\overset`](https://tex.stackexchange.com/a/39230/34974)) would work already via the "big operator limits" layouting when specified using `<mover>`/`<munder>`?

As to the stretchable glyphs/accents, if you're not planning to work on that for now, would it be possible you'd try to outline the thought process and general idea how to try approaching this, as much as you could? Any hints here could be helpful! ❤️ I admit I currently would have no idea how to start neither conceptually, nor where to find the "pieces that are there"... if you wouldn't mind helping me with some hints, maybe I could try experimenting with that!

Thanks for the outline on `<mtext>`; sounds like something that could be doable for me, assuming I can find the SILE functions you mention that could help with that... actually, there might even be something like that in my old code that I tried to write for #220; though not 100% sure; if not, I can maybe try asking @alerque if he'd have some hints.

**Comment by OlivierNicole** [comment]
Tags munder, mover and munderover implemented. See #1240.



> Do you think things equivalent to `\stackrel` ([a.k.a. `\underset`/`\overset`](https://tex.stackexchange.com/a/39230/34974)) would work already via the "big operator limits" layouting when specified using `<mover>`/`<munder>`?

Yes, as long as stretchable operators are part of the equation (pun intended), it should work.

**Comment by OlivierNicole** [comment]
> As to the stretchable glyphs/accents, if you're not planning to work on that for now, would it be possible you'd try to outline the thought process and general idea how to try approaching this, as much as you could? Any hints here could be helpful! heart I admit I currently would have no idea how to start neither conceptually, nor where to find the "pieces that are there"... if you wouldn't mind helping me with some hints, maybe I could try experimenting with that!

When I said “the pieces are there”, I was thinking about the glyphs that were drawn by the OpenType font designers. In the end, the only thing that the math package has to do is positioning these glyphs at the right positions and give them the right sizes to typeset a formula.

Stretchable glyphs are a bit more trickier. They can be done in two ways. First, some glyphs have several variants of increasing sizes: vertical variants (like vertical braces or parentheses) or horizontal variants (like arrows or horizontal braces). When the biggest variant is still too short, arbitrarily long versions can be constructed by overlapping base components of the glyph. The component glyphs, as well as some numeric parameters governing the scaffolding construction, are provided by the font in the MathGlyphConstruction table (see [OpenType spec](https://docs.microsoft.com/en-us/typography/opentype/spec/math)). The parsing of OpenType MATH tables into Lua tables has already been implemented by @harrysummer, so what's left is actually constructing stretchable glyphs, when needed.

I implemented the easy part: for delimiter glyphs, using the vertical variant closest to the target height. I didn't implement anything for horizontal variants, however.

[OpenType Math Illuminated](https://www.tug.org/~vieth/papers/bachotex2009/ot-math-paper.pdf) is probably also a useful reference for the parameters governing the positioning of glyphs.

**Comment by akavel** [comment]
@OlivierNicole thank you so much for the *amazing* description and references :heart:, this will most certainly be helpful to anyone interested in trying to implement this! (which might or might not be me - it makes no sense for me to give any promises here, given my poor track record with delivering stuff related to this issue...)

**Comment by simoncozens** [comment]
This is what a glyph construction table (in Rust) looks like:

```rust
let integral = MathGlyphConstruction {
            glyphAssembly: Offset16::to(GlyphAssembly {
                italicsCorrection: MathValueRecord::new(0),
                partRecords: vec![
                    GlyphPartRecord {
                        glyphID: 2048,
                        startConnectorLength: 0,
                        endConnectorLength: 212,
                        fullAdvance: 893,
                        partFlags: 0,
                    },
                    GlyphPartRecord {
                        glyphID: 2075,
                        startConnectorLength: 893,
                        endConnectorLength: 893,
                        fullAdvance: 893,
                        partFlags: 1,
                    },
                    GlyphPartRecord {
                        glyphID: 2047,
                        startConnectorLength: 212,
                        endConnectorLength: 0,
                        fullAdvance: 893,
                        partFlags: 0,
                    },
                ],
            }),
            mathGlyphVariantRecord: vec![],
        };
```

For the integral sign, there are no math variants, but there is a glyph assembly table, which tells you how to build a stretchable integral sign: start by picking GID 2048 (integralbottom), and then, at anywhere from (top - 212 units) to the top of that glyph, you can place a connector (GID 2075, integral extension). This can be repeated if necessary (`partflags: 1`), and you can cut off up to 893 units from the bottom or top of that glyph to fit the desired height. Finally you take GID 2047 (integraltop) and you can cut off anything from 0 to 212 units from the bottom of that glyph and stick it on top of your extension.

**Comment by OlivierNicole** [comment]
Thanks @simoncozens for this example, it's actually quite illuminating!

**Comment by khaledhosny** [comment]
> For the integral sign, there are no math variants, but there is a glyph assembly table

That is quite odd, I’d expect it to be the other way around for integral signs, what font is that?

**Comment by simoncozens** [comment]
Libertinus.

**Comment by khaledhosny** [comment]
Either your version of the font is broken or the code dumping the MATH table, this is what I see in TTX dump:
```
      <VertGlyphConstruction index="14">
        <!-- VariantCount=2 -->
        <MathGlyphVariantRecord index="0">
          <VariantGlyph value="integral"/>
          <AdvanceMeasurement value="1100"/>
        </MathGlyphVariantRecord>
        <MathGlyphVariantRecord index="1">
          <VariantGlyph value="integral.size1"/>
          <AdvanceMeasurement value="2185"/>
        </MathGlyphVariantRecord>
      </VertGlyphConstruction>
```

**Comment by simoncozens** [comment]
Sorry, not Libertin*us*, LinLibertine_R.otf.

**Comment by khaledhosny** [comment]
I see, makes sense as that fonts MATH table was not that usable.

**Comment by ctrlcctrlv** [comment]
This should be closed.
